src/security/blocking.cpp

```diff
@@ -0,0 +1,409 @@
+    }
+    return ContentType::JSON;
+  }
+};
+
+const std::string_view kDefaultTemplateHtml{
+    "<!DOCTYPE html><html lang=\"en\"><head><meta charset=\"UTF-8\"><meta "
+    "name=\"viewport\" "
+    "content=\"width=device-width,initial-scale=1\"><title>You've been "
+    "blocked</"
+    "title><style>a,body,div,html,span{margin:0;padding:0;border:0;font-size:"
+    "100%;font:inherit;vertical-align:baseline}body{background:-webkit-radial-"
+    "gradient(26% 19%,circle,#fff,#f4f7f9);background:radial-gradient(circle "
+    "at 26% "
+    "19%,#fff,#f4f7f9);display:-webkit-box;display:-ms-flexbox;display:flex;-"
+    "webkit-box-pack:center;-ms-flex-pack:center;justify-content:center;-"
+    "webkit-box-align:center;-ms-flex-align:center;align-items:center;-ms-flex-"
+    "line-pack:center;align-content:center;width:100%;min-height:100vh;line-"
+    "height:1;flex-direction:column}p{display:block}main{text-align:center;"
+    "flex:1;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-"
+    "pack:center;-ms-flex-pack:center;justify-content:center;-webkit-box-align:"
+    "center;-ms-flex-align:center;align-items:center;-ms-flex-line-pack:center;"
+    "align-content:center;flex-direction:column}p{font-size:18px;line-height:"
+    "normal;color:#646464;font-family:sans-serif;font-weight:400}a{color:#"
+    "4842b7}footer{width:100%;text-align:center}footer "
+    "p{font-size:16px}</style></head><body><main><p>Sorry, you cannot access "
+    "this page. Please contact the customer service "
+    "team.</p></main><footer><p>Security provided by <a "
+    "href=\"https://www.datadoghq.com/product/security-platform/"
+    "application-security-monitoring/\" "
+    "target=\"_blank\">Datadog</a></p></footer></body></html>"sv};
+
+const std::string_view kDefaultTemplateJson{
+    "{\"errors\":[{\"title\":\"You've been blocked\",\"detail\":\"Sorry, you "
+    "cannot access this page. Please contact the customer service team. "
+    "Security provided by Datadog.\"}]}"sv};
+}  // namespace
+
+namespace datadog::nginx::security {
+
+// NOLINTNEXTLINE
+std::unique_ptr<BlockingService> BlockingService::instance;
+
+void BlockingService::initialize(std::optional<std::string_view> templ_html,
+                                 std::optional<std::string_view> templ_json) {
+  if (instance) {
+    throw std::runtime_error("Blocking service already initialized");
+  }
+  instance = std::unique_ptr<BlockingService>(
+      new BlockingService(templ_html, templ_json));
+}
+
+void BlockingService::block(BlockSpecification spec, ngx_http_request_t &req) {
+  BlockResponse const resp = BlockResponse::resolve_content_type(spec, req);
+  ngx_str_t *templ{};
+  if (resp.ct == BlockResponse::ContentType::HTML) {
+    templ = &templ_html_;
+  } else if (resp.ct == BlockResponse::ContentType::JSON) {
+    templ = &templ_json_;
+  } else {
+    req.header_only = 1;
+  }
+
+  ngx_http_discard_request_body(&req);
+
+  // TODO: clear all current headers?
```

Comment:
This is important if feasible...

Reply:
Because we don't block on response, the problem becomes moot.